### ISSUE TITLE
[Suggestion] New BIOS Naming System

### ISSUE THREAD

**Issue by LodanZark** [libretro/beetle-psx-libretro#238]
Problem:
========
Usually the bios naming its based in the psx model and not the version of the bios, which can result in the follow situations:
The model SCPH-102 has the European bios 4.4 or 4.5
The american bios 4.1 was installed in the models: SCPH-7001/SCPH-7501/SCPH-7503/SCPH-9001

Solution:
=========
Using the following BIOS naming based in the region and version of it.
If it's trouble some to enforce it, maybe could be done through the hash checksum instead the filename.

--------------------------------------------------

MISC
====

filename:DTLH2000.bin
models:DTLH2000(PlayStation 1 Dev)
crc:18d0f7d8
md5:2118230527a9f51bd9216e32fa912842
sha1:1a8d6f9453111b1d317bb7dae300495fbf54600c

filename:ps2-50j.bin
models: PSX BIOS found on PS2 (NTSC-J)
crc:0dcce9d7
md5:d8f485717a5237285e4d7c5f881b7f32
sha1:d7d6be084f51354bc951d8fa2d8d912aa70abc5e

filename:ps1_rom.bin
models:PSX BIOS found on PS3
crc:2f53b852
md5:81bbe60ba7a3d1cea1d48c14cbcc647b
sha1:c40146361eb8cf670b19fdc9759190257803cab7

--------------------------------------------------

NTSC-UC
=======

ps-20a.bin
DTL-H1001 (Version 2.0 05/07/95 A)
crc:55847d8c
md5:dc2b9bf8da62ec93e868cfd29f0d067d
sha1:649895efd79d14790eabb362e94eb0622093dfb9

ps-21a.bin
DTL-H1101 (Version 2.1 07/17/95 A)
crc:aff00f2f
md5:da27e8b6dab242d8f91a9b25d80c63b8
sha1:ca7af30b50d9756cbd764640126c454cff658479

ps-22a.bin
SCPH-1001/DTL-H1201/DTL-H3001 (Version 2.2 12/04/95 A)
crc:37157331
md5:924e392ed05558ffdb115408c263dccf
sha1:10155d8d6e6e832d6ea66db9bc098321fb5e8ebf

ps-30a.bin
SCPH-5501/SCPH-7003 (Version 3.0 11/18/96 A)
crc:8d8cb7e4
md5:490f666e1afb15b7362b406ed1cea246
sha1:0555c6fae8906f3f09baf5988f00e55f88e9f30b

ps-41a.bin
SCPH-7001/SCPH-7501/SCPH-7503/SCPH-9001 (Version 4.1 12/16/97 A)
crc:502224b6
md5:1e68c231d0896b7eadcad1d7d8e76129
sha1:14df4f6c1e367ce097c11deae21566b4fe5647a9

psone-45a.bin
SCPH-101 (Version 4.5 05/25/00 A)
crc:171bdcec
md5:6e3735ff4c7dc899ee98981385f6f3d0
sha1:dcffe16bd90a723499ad46c641424981338d8378

--------------------------------------------------

NTSC-J
======

filename:ps-10j.bin
models:SCPH-1000/DTL-H1000 (Version 1.0 J)
crc:3b601fc8
md5:239665b1a3dade1b5a52c06338011044
sha1:343883a7b555646da8cee54aadd2795b6e7dd070

filename:ps-11j.bin
models:SCPH-3000/DTL-H1000H (Version 1.1 01/22/95)
crc:3539def6
md5:849515939161e62f6b866f6853006780
sha1:b06f4a861f74270be819aa2a07db8d0563a7cc4e

filename:ps-21j.bin
models:SCPH-3500 (Version 2.1 07/17/95 J)
crc:bc190209
md5:cba733ceeff5aef5c32254f1d617fa62
sha1:e38466a4ba8005fba7e9e3c7b9efeba7205bee3f

filename:ps-22d.bin
models:DTL-H1100 (Version 2.2 03/06/96 D)
crc:decb22f5
md5:ca5cfc321f916756e3f0effbfaeba13b
sha1:73107d468fc7cb1d2c5b18b269715dd889ecef06

filename:ps-22j.bin
models:SCPH-5000/DTL-H1200 (Version 2.2 12/04/95 J)
crc:24fc7e17
md5:57a06303dfa9cf9351222dfcbb4a29d9
sha1:ffa7f9a7fb19d773a0c3985a541c8e5623d2c30d

filename:ps-22j(v).bin
models: SCPH-5903 (VideoCD support)
crc:446ec5b2
md5:81328b966e6dcf7ea1e32e55e1c104bb
sha1:15c94da3cc5a38a582429575af4198c487fe893c

ps-30j.bin
SCPH-5500 (Version 3.0 09/09/96 J)
crc:ff3eeb8c
md5:8dd7d5296a650fac7319bce665a6a53c
sha1:b05def971d8ec59f346f2d9ac21fb742e3eb6917

ps-40j.bin
SCPH-7000/SCPH-9000 (Version 4.0 08/18/97 J)
crc:ec541cd0
md5:8e4c14f567745eff2f0408c8129f72a6
sha1:77b10118d21ac7ffa9b35f9c4fd814da240eb3e9

filename:ps-41a(w).bin
models:SCPH-7000W (Playstation 10 Million Model, BIOS based on NTSC-UC, **REGION FREE**)
crc:b7c43dad
md5:b84be139db3ee6cbd075630aa20a6553
sha1:1b0dbdb23da9dc0776aac58d0755dc80fea20975

psone-43j.bin
SCPH-100 (Version 4.3 03/11/00 J)
crc:f2af798b
md5:8abc1b549a4a80954addc48ef02c4521
sha1:339a48f4fcf63e10b5b867b8c93cfd40945faf6c

--------------------------------------------------

PAL-E
=====

ps-20e.bin
DTL-H1002/SCPH-1002 (Version 2.0 05/10/95 E)
crc:9bb87c4b
md5:54847e693405ffeb0359c6287434cbef
sha1:20b98f3d80f11cbf5a7bfd0779b0e63760ecc62c

ps-21e.bin
SCPH-1002/DTL-H1102 (Version 2.1 07/17/95 E)
crc:86c30531
md5:417b34706319da7cf001e76e40136c23
sha1:76cf6b1b2a7c571a6ad07f2bac0db6cd8f71e2cc

ps-22e.bin
SCPH-1002/DTL-H1202/DTL-H3002 (Version 2.2 12/04/95 E)
crc:1e26792f
md5:e2110b8a2b97a8e0b857a45d32f7e187
sha1:b6a11579caef3875504fcf3831b8e3922746df2c

ps-30e.bin
SCPH-5502/SCPH-5552 (Version 3.0 01/06/97 E)
crc:d786f0b9
md5:32736f17079d0b2b7024407c39bd3050
sha1:f6bc2d1f5eb6593de7d089c425ac681d6fffd3f0

ps-41e.bin
SCPH-7002/SCPH-7502/SCPH-9002 (Version 4.1 12/16/97 E)
crc:318178bf
md5:b9d9a0286c33dc6b7237bb13cd46fdee
sha1:8d5de56a79954f29e9006929ba3fed9b6a418c1d

psone-44e.bin
SCPH-102 (Version 4.4 03/24/00 E)
crc:0bad7ea9
md5:b10f5e0e3d9eb60e5159690680b1e774
sha1:beb0ac693c0dc26daf5665b3314db81480fa5c7c

psone-45e.bin
SCPH-102 (Version 4.5 05/25/00 E)
crc:76b880e5
md5:de93caec13d1a141a40a79f5c86168d6
sha1:dbc7339e5d85827c095764fc077b41f78fd2ecae

--------------------------------------------------

**Comment by simias** [comment]
I use checksum-based BIOS detection in Rustation, I find it quite convenient. At least you don't have to worry about getting the name exactly right, you just dump your BIOS into the system folder and you're good to go. And as you mention that way there's no ambiguity about which exact BIOS you're using.

I don't think changing the file name is a good idea, in practice very few people dump their own BIOSes so they end up downloading them from some random internet website and they almost all use the SCPH-xxxx naming scheme. Asking the user to rename their BIOS files when they might not even know which version it is exactly is simply not friendly.

Here's the database I use in Rustation: https://github.com/simias/rustation/blob/master/src/bios/db.rs

**Comment by rz5** [comment]
@simias - I had an idea of how to make BIOS detection more versatile and wrote it in issue #239. It boiled down But in hindsight, it seems your way would be even more versatile but slower.

My 'system' directory has about 60 files on the top layer. Worst case scenario is we do 59 checksums till we arrive on the right BIOS file.

To reduce this overhead, how about we do what we did for the redream and ppsspp cores and require that the user create a folder inside the 'system' dir, called e.g. 'psx' and put his BIOS dumps there?

@fr500 @twinaphex, thoughts? Of course, after this change is done, it would have to be advertised loudly so the user isn't caught by surprise.

**Comment by retro-wertz** [comment]
question(this does not mean i disagree with any updates possible-as alot of bios floating around are just renamed ones), aside from us, eu, jp bios sets, are there specific games requiring later revisions of bios?

i do agree about matching the bios based on crc/md5 though as opposed to by filename(just incase someone was just renaming stuff).

thanks

**Comment by simias** [comment]
@rz5 In order not to checksum all the files in the system folder I first check the size of the file, BIOS images are always exactly 512KB (512 * 1024 bytes). That's a very quick test and should get rid of most, if not all, non-PSX BIOS files.

And even if there are a couple of false positives hashing 512KB shouldn't be noticeably long on any system capable of running beetle in the first place. In Rustation I've used this system with the big retroarch firmware/bios/whatever dump that's floating around and never noticed any delay.

**Comment by LodanZark** [comment]
I added the sha256 hashs just in case

**Comment by simias** [comment]
Yeah I don't think there's any reason to use MD5 or SHA-1 on a new application these days.

**Comment by deltabeard** [comment]
MD5 is [much faster](https://stackoverflow.com/a/2723941) than SHA-#, and since this isn't a security related operation, I don't what the issue would be with using MD5 in this case.
Of course the file size of a BIOS image is small anyway, so the difference in speed may not be noticeable.

**Comment by simias** [comment]
Yeah, speed isn't a factor here and if it was there are other hash algorithms that run much faster than SHA-256 without having the flaws of MD5 and SHA-1. I mean, if you really don't care about security you might as well go with CRC-32.

That being said it's not security related until it is, what if somebody decides to craft a broken BIOS for fun? Then you end up with bug reports about your emulator not working and good luck figuring that out.

Unlikely? Sure. But why even bother with that when there are better alternatives?

MD5 and SHA-1 are dead hash functions, it's time to let them rest in peace.

**Comment by rz5** [comment]
@LodanZark - Would you consider pitching this idea to Ryphecha? If she adopts it, it'll make more sense for this fork to implement it too.

**Comment by LodanZark** [comment]
@rz5 Isn't she the main dev of mednafen? I don't know how to contact her, does mednafen have github page?

**Comment by rz5** [comment]
@LodanZark Yes she is. You can reach her via IRC; #mednafen channel over at freenode.
